src/voice/VoiceRecorder.ts

```diff
@@ -0,0 +1,113 @@
+/*
+Copyright 2021 The Matrix.org Foundation C.I.C.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+import * as Recorder from 'opus-recorder';
+import encoderPath from 'opus-recorder/dist/encoderWorker.min.js';
+import {MatrixClient} from "matrix-js-sdk/src/client";
+import CallMediaHandler from "../CallMediaHandler";
+import {SimpleObservable} from "matrix-widget-api";
+
+export class VoiceRecorder {
+    private recorder = new Recorder({
+        encoderPath, // magic from webpack
+        mediaTrackConstraints: <MediaTrackConstraints>{
+            deviceId: CallMediaHandler.getAudioInput(),
+        },
+        encoderSampleRate: 16000, // we could go down to 12khz, but we lose quality
```

Comment:
12kHz would be pretty weird. Honestly I would probably just go up to 48Khz: it's what browsers will use over WebRTC.

Reply:
Also I'd expect to see a mono / stereo param here maybe (unless that's implicit from the source?) And more importantly a target bitrate / quality setting?

Reply:
apaprently it defaults to mono, and picks a quality/bitrate for us. I'll play around with some options that seem to fit well for us and hardcode them for good measure.

For the sample rate: 16khz didn't distort my voice at all - it was indistinguishable to 48khz, and ultimately lowered file size. The file size gains aren't super large though (at 16khz it's about 34kb for 5s of audio, at 48khz it's 45kb for 5s).